### ISSUE TITLE
Usage with node-forge via streamBuilder

### ISSUE THREAD

**Issue by Aareksio** [mqttjs/MQTT.js#949]
Hey,

first of all, I am aware this may not be the appropriate place to ask the question, but it is the best place I found.

---

### **TLDR**
How to supply node-forge (or other non-stream) TLS connection to MQTT.js
_or more broadly_
How to use HSM to secure MQTT TLS connection

### **Problem**
I want to delegate TLS signing to Hardware Security Module (HSM), precisely ATECC608A, in order to securely connect to AWS IoT MQTT broker.

Doing so using native TLS connection would, to my current knowledge, require low level code and openSSL configuration / compilation - after searching the most prominent solution seems to be using node-forge patched for TLS 1.2 support (which my project requires).

The issue with node-forge is that it's TCP connection and socket are not compatible with node.js net.Socket, hence it's impossible to pass them to mqtt.js directly. I tried multiple methods to wrap it around net.Socket and stream.Duplex, without luck.

I tried using PKCS11 OpenSSL engine and crypto.setEngine, no luck there as well.

### **Prior art** _(that I'm aware of)_

[node-atecc](https://github.com/wisnuc/node-atecc) and [telsa](https://github.com/wisnuc/telsa) - the project description sound promising, but I have trouble trusting connection security to a project reading the description more closely and checking commit history - I'll be sure to revisit it soon though.

[node-https-hsm](https://github.com/githoniel/node-https-hsm) - it does pretty much what I need, uses forge.tls to encrypt net.Socket, but I could figure out how to tailor it for mqtt.js needs exactly

---

I am aware this may be niche use case and trying to solve the problem in Node is not the best approach, but at the moment I lack the knowledge required to maneuver in C++ code of OpenSSL (or alternatives, such as WolfSSL). 

All and any input is welcome.

**Comment by heol-files** [comment]
Hi, have you solved this?
I Do it with openssl in "normal" node JS, but when using Electron this does not work, because Electron does not use openssl crypto engine (it uses boringssl) :-(
So I wounder if you found a way of doing this in forge?
